base/lib/src/storage/task/put_parts_task/upload_parts_task.dart

```diff
@@ -38,118 +49,173 @@ class UploadPartsTask extends RequestTask<List<Part>> with CacheMixin {
-    this.host,
-    this.partSize,
-    this.uploadId,
-    this.file,
-    this.bucket,
+    @required this.file,
+    @required this.token,
+    @required this.bucket,
+    @required this.uploadId,
+    @required this.host,
+    @required this.partSize,
+    @required this.maxPartsRequestNumber,
     this.key,
-    this.maxPartsRequestNumber,
   });
 
+  static String getCacheKey(
+    String path,
+    int length,
+    int partSize,
+    String key,
+  ) {
+    final keyMap = {
+      'key': key,
+      'path': path,
+      'file_size': length,
+      'part_size': partSize,
+    };
+
+    final keyString =
+        keyMap.entries.map((e) => '${e.key}/${e.value}').join('/');
+    return 'qiniu_dart_sdk_upload_parts_task@[$keyString]';
+  }
+
   @override
   void preStart() {
     _fileByteLength = file.lengthSync();
     _partByteLength = partSize * 1024 * 1024;
     _idleRequestNumber = maxPartsRequestNumber;
-    _cacheKey =
-        'qiniu_dart_sdk_upload_parts_task_${file.path}_key_${key}_size_$_fileByteLength';
+    _totalPartCount = (_fileByteLength / _partByteLength).ceil();
+    _cacheKey = getCacheKey(file.path, _fileByteLength, partSize, key);
+    recoverUploadedPart();
     super.preStart();
   }
 
   @override
   void postReceive(data) {
-    setCache(json.encode(data));
+    storeUploadPart();
     super.postReceive(data);
   }
 
   @override
-  void postError(error) {
+  void postError(Object error) {
     /// 取消，网络问题等可能导致上传中断，缓存已上传的分片信息
-    setCache(json.encode(_parts));
+    storeUploadPart();
     super.postError(error);
   }
 
-  @override
-  Future<List<Part>> createTask() async {
-    final uploadParts = getCache();
-    if (uploadParts != null) {
-      _parts.addAll(json.decode(uploadParts) as List<Part>);
+  void storeUploadPart() {
+    if (_uploadPartMap.isEmpty) {
+      return;
     }
-    final com = Completer<List<Part>>();
-    _uploadParts(() => com.complete(_parts), com.completeError);
 
-    return com.future;
+    setCache(jsonEncode(_uploadPartMap.values.toList()));
   }
 
-  void _uploadParts(void Function() done, void Function(dynamic) error) {
-    while (_idleRequestNumber > 0 && _byteStartOffset < _fileByteLength) {
-      _partNumber++;
+  // 从缓存恢复已经上传的 part
+  void recoverUploadedPart() {
+    /// 获取缓存
+    final cachedData = getCache();
+
+    /// 尝试从缓存恢复
+    if (cachedData != null) {
+      var cachedList = <Part>[];
+
+      try {
+        cachedList = json.decode(cachedData) as List<Part>;
+      } catch (error) {
+        //
+      }
+
+      for (final part in cachedList) {
+        _uploadPartMap[part.partNumber] = part;
+      }
+    }
+  }
+
+  @override
+  Future<List<Part>> createTask() async {
+    await _uploadParts();
+    return _uploadPartMap.values.toList();
+  }
 
-      final cachedPart = _parts.isNotEmpty
-          ? _parts.firstWhere((element) => element.partNumber == _partNumber)
-          : null;
+  Future<void> _uploadParts({int startPartNumber = 1}) async {
```

Comment:
```suggestion
  Future<void> _uploadParts(int startPartNumber) async {
```

建议不要有默认值，不希望除了开头有默认的情况的话，就开头那里传个 1 就行了